test/stdlib/Result.swift

```diff
@@ -0,0 +1,171 @@
+    case let .success(success):
+      return success
+    case .failure:
+      return nil
+    }
+  }
+  
+  var failure: Failure? {
+    switch self {
+    case .success:
+      return nil
+    case let .failure(failure):
+      return failure
+    }
+  }
+}
+
+ResultTests.test("Construction") {
+  let result1: Result<String, Err> = .success(string)
+  let result2: Result<String, Err> = .failure(.err)
+  let string1: String? = {
+    switch result1 {
+      case let .success(string): 
+        return string
+      case .failure: 
+        expectUnreachable()
+        return nil
+    }
+  }()
+  let error: Err? = {
+    switch result2 {
+      case let .failure(failure): 
+        return failure
+      case .success: 
+        expectUnreachable()
+        return nil
+    }
+  }()
+
+  expectEqual(string1, string)
+  expectEqual(error, .err)
+}
+
+ResultTests.test("Throwing Initialization and Unwrapping") {
+  func notThrowing() throws -> String {
+    return string
+  }
+
+  func throwing() throws -> String {
+    throw Err.err
+  }
+    
+  let result1 = Result { try throwing() }
+  let result2 = Result { try notThrowing() }
+  
+  expectEqual(result1.failure as? Err, Err.err)
+  expectEqual(result2.success, string)
+    
+  do {
+    _ = try result1.get()
+  } catch let error as Err {
+    expectEqual(error, Err.err)
+  } catch {
+    expectUnreachable()
+  }
+    
+  do {
+    let unwrapped = try result2.get()
+    expectEqual(unwrapped, string)
+  } catch {
+    expectUnreachable()
+  }
+    
+  // Test unwrapping strongly typed error.
+  let result3 = Result<String, Err>.failure(Err.err)
+  do {
+    _ = try result3.get()
+  } catch let error as Err {
+    expectEqual(error, Err.err)
+  } catch {
+    expectUnreachable()
+  }
+}
+
+ResultTests.test("Functional Transforms") {
+  func transformDouble(_ int: Int) -> Int {
+    return 2 * int
+  }
+  
+  func transformTriple(_ int: Int) -> Int {
+    return 3 * int
+  }
+  
+  func transformError(_ err: Err) -> Err {
+    if err == .err {
+      return .derr
+    } else {
+      return .err
+    }
+  }
+
+  func resultValueTransform(_ int: Int) -> Result<Int, Err> {
+    return .success(transformDouble(int))
+  }
+  
+  func resultErrorTransform(_ err: Err) -> Result<Int, Err> {
+    return .failure(transformError(err))
+  }
+    
+  let result1: Result<Int, Err> = .success(1)
+  let newResult1 = result1.map(transformDouble)
+    
+  expectEqual(newResult1, .success(2))
+    
+  let result2: Result<Int, Err> = .failure(.err)
+  let newResult2 = result2.mapError(transformError)
+    
+  expectEqual(newResult2, .failure(.derr))
+    
+  let result3: Result<Int, Err> = .success(1)
+  let newResult3 = result3.flatMap(resultValueTransform)
+    
+  expectEqual(newResult3, .success(2))
+    
+  let result4: Result<Int, Err> = .failure(.derr)
+  let newResult4 = result4.flatMapError(resultErrorTransform)
+    
+  expectEqual(newResult4, .failure(.err))
+}
+
+ResultTests.test("Equatable") {
```

Comment:
Did you consider using the `checkEquatable` from StdlibUnitTest?

Reply:
No, mostly cribbed these tests by looking at those from `Optional` and earlier feedback. I didn't know that existed. Can you provide an example of what they would simplify here?

In general, is there any documentation around this special `StdlibUnittest` API or testing requirements for the standard library?

Reply:
I was just thinking out loud really. These conformances are synthesized by the compiler, and should be working. Otherwise we're in trouble.

> In general, is there any documentation around this special StdlibUnittest API or testing requirements for the standard library?
Unfortunately no.

Reply:
> Can you provide an example of what they would simplify here?
It's not about simplification. Just that `checkEquatable` tests for the cases you and I might not necessarily think about, like whether `==` is transitive or not, or whether the implementation is correct in that `==` is the opposite of `!=`.

Reply:
I've added a use of `checkEquatable`, in additional to the tests that were already there. Let me know if there're any other scenarios you want tested, or if you feel the tests are now redundant.

Reply:
Note: `checkHashable` starts by calling `checkEquatable` -- so it is okay to omit the `checkEquatable` test if we have tests elsewhere with `checkHashable`.